tiatoolbox/models/engine/engine_abc.py

```diff
@@ -51,92 +63,193 @@ def prepare_engines_save_dir(
-            Number of workers to postprocess the results of the model. default = 0
         device (str):
-            Select the device to run the model. Default is "cpu".
+            Select the device to run the model. Please see
+            https://pytorch.org/docs/stable/tensor_attributes.html#torch.device
+            for more details on input parameters for device. Default is "cpu".
         verbose (bool):
-            Whether to output logging information.
+            Whether to output logging information. Default value is False.
 
     Attributes:
         images (list of str or list of :obj:`Path` or NHWC :obj:`numpy.ndarray`):
-            A NHWC image or a path to WSI.
+            A list of image patches in NHWC format as a numpy array
+            or a list of str/paths to WSIs.
+        masks (list of str or list of :obj:`Path` or NHWC :obj:`numpy.ndarray`):
+            A list of tissue masks or binary masks corresponding to processing area of
+            input images. These can be a list of numpy arrays or paths to
+            the saved image masks. These are only utilized when patch_mode is False.
+            Patches are only generated within a masked area.
+            If not provided, then a tissue mask will be automatically
+            generated for whole slide images.
         patch_mode (str):
-            Whether to treat input image as a patch or WSI.
-            default = True.
-        model (str | nn.Module):
-            Defined PyTorch model.
-            Name of an existing model supported by the TIAToolbox for
-            processing the data. For a full list of pretrained models,
+            Whether to treat input images as a set of image patches. TIAToolbox defines
+            an image as a patch if HWC of the input image matches with the HWC expected
+            by the model. If HWC of the input image does not match with the HWC expected
+            by the model, then the patch_mode must be set to False which will allow the
+            engine to extract patches from the input image.
+            In this case, when the patch_mode is False the input images are treated
+            as WSIs. Default value is True.
+        model (str | ModelABC):
+            A PyTorch model or a name of an existing model from the TIAToolbox model zoo
+            for processing the data. For a full list of pretrained models,
             refer to the `docs
             <https://tia-toolbox.readthedocs.io/en/latest/pretrained.html>`_
             By default, the corresponding pretrained weights will also
             be downloaded. However, you can override with your own set
             of weights via the `weights` argument. Argument
             is case-insensitive.
         ioconfig (ModelIOConfigABC):
-            Input IO configuration to run the Engine.
-        _ioconfig ():
+            Input IO configuration of type :class:`ModelIOConfigABC` to run the Engine.
+        _ioconfig (ModelIOConfigABC):
```

Comment:
Do we really need two different ioconfigs? I think it is a source of potential confusion to users. If we really need both and can't find a way to have just one, we at least should provide more info about what distinguishes them, why they both are needed and how they interact. For example I think a user might wonder things like, if I update ioconfig will runtime _ioconfig automatically update to match?

Reply:
Hovernet requires these two ioconfigs. We can revisit this when we finalise engines for segmentation models.